lib/cdo/crowdin/utils.rb

```diff
@@ -0,0 +1,90 @@
+require 'json'
+require 'parallel'
+
+module Crowdin
+  class Utils
+    # @param project [Crowdin::Project]
+    # @param options [Hash, nil]
+    # @param options.changes_json [String, nil] path to file where files with
+    #  changes will be written out in JSON format
+    # @param options.etags_json [String, nil] path to file where etags will be
+    #  written out in JSON format
+    # @param options.locales_dir [String, nil] path to directory where changed
+    #  files should be downloaded
+    # @param options.logger [Logger, nil]
+    def initialize(project, options={})
+      @project = project
+      @changes_json = options.fetch(:changes_json, "/tmp/#{project.id}_changes.json")
+      @etags_json = options.fetch(:etags_json, "/tmp/#{project.id}_etags.json")
+      @locales_dir = options.fetch(:locales_dir, "/tmp/locales")
+      @logger = options.fetch(:logger, Logger.new(STDOUT))
+    end
+
+    # Fetch from Crowdin a list of files changed since the last sync. Uses
+    # etags sourced from the @etags_json file to define what we mean by "since
+    # the last sync," and writes the results out to @changes_json.
+    def fetch_changes
+      etags = File.exist?(@etags_json) ? JSON.parse(File.read(@etags_json)) : {}
+      changes = {}
+
+      languages = @project.languages
+      num_languages = languages.length
+      languages.each_with_index do |language, i|
+        language_code = language["code"]
+        @logger.debug("#{language['name']} (#{language_code}): #{i}/#{num_languages}")
+        @logger.info("~#{(i * 100 / num_languages).round(-1)}% complete (#{i}/#{num_languages})") if i > 0 && i % (num_languages / 5) == 0
+
+        etags[language_code] ||= {}
+        files = @project.list_files
+
+        changed_files = Parallel.map(files) do |file|
+          etag = etags[language_code].fetch(file, nil)
+          response = @project.export_file(file, language_code, etag: etag, only_head: true)
+          case response.code
+          when 200
+            [file, response.headers["etag"]]
+          when 304
+            nil
+          else
+            raise "cannot handle response code #{response.code}"
+          end
+        end.compact
+
+        next if changed_files.empty?
+
+        changes[language_code] = changed_files.to_h
+        etags[language_code].merge!(changes[language_code])
+        File.write(@etags_json, JSON.pretty_generate(etags))
+        File.write(@changes_json, JSON.pretty_generate(changes))
+      end
+
+      # Write changes out one final time; in the case that no language had
+      # changes, we would otherwise skip the write on every loop and leave any
+      # existing changes file unchanged. We could alternatively achieve this by
+      # writing (or removing) the file before starting.
+      File.write(@changes_json, JSON.pretty_generate(changes))
+    end
+
+    # Downloads all files referenced in @changes_json to @locales_dir
+    def download_changed_files
```

Comment:
Does it make sense to run `download_changed_files` without running `fetch_changes` first? Should there be an assert or warning message logged so future developers don't break this relationship?

Reply:
I'm picturing something like a situation where the sync out goes sideways, so you want to just throw everything away and start over. In that situation, you could just rerun `download_changed_files` without actually needing to rerun `fetch_changes`. But that does assume the changes.json is still around from the previous run; I'll add an explicit error message for the situation where there is no `changes.json` to be found